compiler/rustc_mir_transform/src/errors.rs

```diff
@@ -0,0 +1,275 @@
+};
+use rustc_macros::{Diagnostic, LintDiagnostic, Subdiagnostic};
+use rustc_middle::mir::{AssertKind, UnsafetyViolationDetails};
+use rustc_session::lint::{self, Lint};
+use rustc_span::Span;
+
+#[derive(LintDiagnostic)]
+pub(crate) enum ConstMutate {
+    #[diag(mir_transform_const_modify)]
+    #[note]
+    Modify {
+        #[note(mir_transform_const_defined_here)]
+        konst: Span,
+    },
+    #[diag(mir_transform_const_mut_borrow)]
+    #[note]
+    #[note(mir_transform_note2)]
+    MutBorrow {
+        #[note(mir_transform_note3)]
+        method_call: Option<Span>,
+        #[note(mir_transform_const_defined_here)]
+        konst: Span,
+    },
+}
+
+#[derive(Diagnostic)]
+#[diag(mir_transform_unaligned_packed_ref, code = "E0793")]
+#[note]
+#[help]
+pub(crate) struct UnalignedPackedRef {
+    #[primary_span]
+    pub span: Span,
+}
+
+#[derive(LintDiagnostic)]
+#[diag(mir_transform_unused_unsafe)]
+pub(crate) struct UnusedUnsafe {
+    #[label(mir_transform_unused_unsafe)]
+    pub span: Span,
+    #[label]
+    pub nested_parent: Option<Span>,
+}
+
+pub(crate) struct RequiresUnsafe {
+    pub span: Span,
+    pub details: RequiresUnsafeDetail,
+    pub enclosing: Option<Span>,
+    pub op_in_unsafe_fn_allowed: bool,
+}
+
+impl<'sess, G: EmissionGuarantee> IntoDiagnostic<'sess, G> for RequiresUnsafe {
+    #[track_caller]
+    fn into_diagnostic(self, handler: &'sess Handler) -> DiagnosticBuilder<'sess, G> {
+        let mut diag =
+            handler.struct_diagnostic(crate::fluent_generated::mir_transform_requires_unsafe);
+        diag.code(rustc_errors::DiagnosticId::Error("E0133".to_string()));
+        diag.set_span(self.span);
+        diag.span_label(self.span, self.details.label());
+        diag.note(self.details.note());
+        let desc = handler.eagerly_translate_to_string(self.details.label(), [].into_iter());
+        diag.set_arg("details", desc);
+        diag.set_arg("op_in_unsafe_fn_allowed", self.op_in_unsafe_fn_allowed);
+        if let Some(sp) = self.enclosing {
+            diag.span_label(sp, crate::fluent_generated::mir_transform_not_inherited);
+        }
+        diag
+    }
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum RequiresUnsafeDetail {
+    CallToUnsafeFunction(Span),
+    UseOfInlineAssembly(Span),
+    InitializingTypeWith(Span),
+    CastOfPointerToInt(Span),
+    UseOfMutableStatic(Span),
+    UseOfExternStatic(Span),
+    DerefOfRawPointer(Span),
+    AccessToUnionField(Span),
+    MutationOfLayoutConstrainedField(Span),
+    BorrowOfLayoutConstrainedField(Span),
+    CallToFunctionWith(Span),
+}
+
+impl RequiresUnsafeDetail {
+    pub fn from_details(details: UnsafetyViolationDetails) -> impl Fn(Span) -> Self {
+        use RequiresUnsafeDetail as B;
+        use UnsafetyViolationDetails as A;
+        match details {
+            A::CallToUnsafeFunction => B::CallToUnsafeFunction,
+            A::UseOfInlineAssembly => B::UseOfInlineAssembly,
```

Comment:
Would it make more sense to turn `RequiresUnsafeDetail` into a struct containing the span and unsafety violation? That should remove this redundant looking function.